File: android/lib/src/main/java/org/silvermine/downloadmanager/DownloadItem.kt
package org.silvermine.downloadmanager

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Emit-only payload sent to the frontend. Built from a [DownloadRecord] with
 * `progress` derived from [receivedBytes] / [totalBytes]; never persisted, so
 * nothing internal to the download machinery leaks into the frontend contract.
 * [totalBytes] is `null` when the server supplied no content length, and
 * serializes as an explicit JSON `null`.
 *
 * No property carries a default, so every key here is written whatever
 * `encodeDefaults` the caller's `Json` uses, rather than depending on the
 * TypeScript layer to coalesce it (attachDownload in `guest-js/actions.ts`).
 * [CreateOptions] carries one but marks it `@Required`, so it is written too.
 */
@Serializable
data class DownloadItem(
   @SerialName("url")
   val url: String,

   @SerialName("path")
   val path: String,

   /** Network policy fixed when this download was created. */
   @SerialName("options")
   val options: CreateOptions,

   @SerialName("receivedBytes")
   val receivedBytes: Long,

   @SerialName("totalBytes")
   val totalBytes: Long?,

   @SerialName("progress")
   val progress: Double,

   @SerialName("status")
   val status: DownloadStatus,
)

File: ios/DownloadManagerKit/Sources/DownloadManagerKit/DownloadItem.swift
//
//  DownloadItem.swift
//  DownloadManagerKit
//

import Foundation

/// Emit-only payload sent to the frontend. Built from a [`DownloadRecord`] with
/// `progress` derived from `receivedBytes` / `totalBytes`; never persisted and
/// never decoded, so internal state such as the resume-data path stays out of
/// the frontend contract. `totalBytes` is nil when the server supplied no
/// content length, and encodes as an explicit JSON `null`.
public struct DownloadItem: Identifiable, Encodable, Sendable {
   public var id: URL { path }
   
   public let url: URL
   public let path: URL
   public let receivedBytes: UInt64
   public let totalBytes: UInt64?
   public let progress: Double
   public let status: DownloadStatus

   init(
      url: URL,
      path: URL,
      receivedBytes: UInt64,
      totalBytes: UInt64?,
      progress: Double,
      status: DownloadStatus
   ) {
      self.url = url
      self.path = path
      self.receivedBytes = receivedBytes
      self.totalBytes = totalBytes
      self.progress = progress
      self.status = status
   }
   
   enum CodingKeys: String, CodingKey {
      case url, path, receivedBytes, totalBytes, progress, status
   }
   
   public func encode(to encoder: Encoder) throws {
      var container = encoder.container(keyedBy: CodingKeys.self)

      try container.encode(url, forKey: .url)
      try container.encode(path, forKey: .path)
      try container.encode(receivedBytes, forKey: .receivedBytes)
      try container.encode(progress, forKey: .progress)
      try container.encode(status, forKey: .status)

      // The synthesized encoding would use encodeIfPresent and omit the key. The
      // TypeScript layer coalesces a missing key to null anyway (attachDownload in
      // guest-js/actions.ts), so this is not load-bearing for the plugin — it keeps
      // the payload self-describing for anything reading DownloadManagerKit directly.
      if let totalBytes = totalBytes {
         try container.encode(totalBytes, forKey: .totalBytes)
      } else {
         try container.encodeNil(forKey: .totalBytes)
      }
   }
}
